[tzuennn] iP

tzuennn opened the pull request:
# Nugget
> "Your mind is for having ideas, not holding them." - David Allen [source](https://dansilvestre.com/productivity-quotes/)

Nugget frees your mind of having to remember things you need to do. It's,
- text-based
- easy to learn
- ~~FAST~~ *SUPER* FAST to use

All you need to do is,
1. download it from [here](https://nus-cs2103-ay2425s1.github.io/website/schedule/week4/project.html).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😄 

And it is **FREE!**

Features:
- [x] Managing tasks
- [ ] Managing deadlines (coming soon!)
- [ ] Reminders (coming soon!)

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

tzuennn committed:
Add Nugget.java

tzuennn committed:
Add bot echo feature

tzuennn committed:
Store and display text by user

tzuennn committed:
Add ability to mark and unmark tasks

tzuennn committed:
Add 3 diff task modes

tzuennn committed:
Add test cases

tzuennn committed:
Error handling of invalid inputs

tzuennn committed:
Add delete function and tests

tzuennn committed:
Add saving feature locally

tzuennn committed:
Merge branch 'branch-Level-7'

tzuennn committed:
Change type of date from String to DateTime

tzuennn committed:
Merge branch 'branch-Level-8'

tzuennn committed:
Change date's type from String to DateTime

tzuennn committed:
Merge branch 'branch-Level-8'

Merge with branch-Level-8

tzuennn committed:
Separate classes, following OOP

tzuennn committed:
Merge branch 'A-MoreOOP'

tzuennn committed:
Use package to organise files

tzuennn committed:
Merge branch 'A-Packages'

tzuennn committed:
Merge branch 'add-gradle-support'

tzuennn committed:
Set up gradle

tzuennn committed:
Add test cases for Parser and TaskList

tzuennn committed:
Merge branch 'A-JUnit'

tzuennn committed:
Add javadoc for most files

tzuennn committed:
Fix format for switch statements

tzuennn committed:
Add find keyword function for task searching

tzuennn committed:
Merge branch 'A-JavaDoc'

tzuennn committed:
Merge branch 'A-CodingStandard'

tzuennn committed:
Resolve conflict in switch statements

jinhanjinhan commented on src/main/java/nugget/Parser.java:
```diff
@@ -0,0 +1,122 @@
+            throw new EmptyDescriptionException();
+        }
+    }
+
```
I like how you have a method specifically for parsing index to check whether the index is valid!

jinhanjinhan commented on src/main/java/nugget/Storage.java:
```diff
@@ -0,0 +1,81 @@
+package nugget;
```
Not sure where to put this,  but I am thinking would it be better to separate the classes to separate sub packages? Maybe something like nugget.exception and nugget.command?

jinhanjinhan commented on src/main/java/nugget/Storage.java:
```diff
@@ -0,0 +1,81 @@
+     * @return An {@code ArrayList} of tasks loaded from the file.
+     */
+    public ArrayList<Task> loadTasks() {
+        ArrayList<Task> tasks = new ArrayList<>();
```
Great naming of "tasks" instead of "task"

jinhanjinhan commented on src/main/java/nugget/TaskList.java:
```diff
@@ -0,0 +1,53 @@
+    }
+
+    public void addTask(Task task) {
+        tasks.add(task);
```
Once again, great differentiation between "task" and "tasks"!

jinhanjinhan commented on src/main/java/nugget/Ui.java:
```diff
@@ -0,0 +1,69 @@
+
+public class Ui {
+
+    public void line() {
```
For this part, I think it would be good if you use printLine() instead of print() to make it in the form of an action

jinhanjinhan commented on src/test/java/nugget/TaskListTest.java:
```diff
@@ -0,0 +1,77 @@
+        assertTrue(taskList.getTask(0).isCompleted(), "Task should be marked as completed");
+    }
+
+    @Test
```
For the test methods, maybe a better name would be in the form of unitBeingTested_descriptionOfTestInputs_expectedOutcome (as specified from the notes)?

Wnayar replied:
I agree with this comment.

tzuennn replied:
Thank you for pointing that out!

jinhanjinhan reviewed (commented):
Overall the naming is great! Most parts follow the naming convention, except for the testing methods.

Wnayar commented on src/main/java/nugget/MarkTaskCommand.java:
```diff
@@ -0,0 +1,35 @@
+     * @param ui The user interface that displays messages to the user.
+     * @param storage The storage that handles the saving of tasks.
+     * @throws NuggetException If the provided task index is invalid.
+     */
```
Nice descriptive header comments for public methods.

Wnayar commented on src/main/java/nugget/Ui.java:
```diff
@@ -0,0 +1,69 @@
+        System.out.println("________________________________________");
+        System.out.println("Hello! I'm nugget.Nugget");
+        System.out.println("What can I do for you?");
+        System.out.println("________________________________________");
```
Awesome looking UI. Have you considered using a variable for the line breaks?

Wnayar commented on src/main/java/nugget/FindTaskCommand.java:
```diff
@@ -0,0 +1,20 @@
+    }
+
+    @Override
+    public void execute(TaskList tasks, Ui ui, Storage storage) throws NuggetException {
```
I think the abstraction of each command is great. Have you considered giving a more descriptive function name, because in the context of FindTaskCommand now its okay because its the only function. But say in the future if you add to this class another function then it may not be as clear what the function purpose is. But overall awesome work!

Wnayar commented on src/main/java/nugget/TaskList.java:
```diff
@@ -0,0 +1,53 @@
+        tasks.get(index).unmarkTask();
+    }
+
+    public int size() {
```
Great naming conventions. I think here size could be replace with something more verb like. Perhaps getSize().

Wnayar reviewed (commented):
LGTM. Just some nits related to CS2103T coding standards to fix.

tzuennn committed:
Implement basic GUI for chatbot

tzuennn committed:
Merge branch 'branch-level-10'

tzuennn committed:
Configure to allow checkstyle

tzuennn committed:
Add assertions in ChatUI and Parser classes

Assertions are missing in critical parts of the code. Adding assertions
helps ensure that certain conditions hold true during runtime, which
can prevent unexpected behaviour and help in debugging.

Include assertions in ChatUI to validate that images are loaded and
present in the specified path. Additional assertions were added in
Parser to verify correctness of user input parsing.

This will catch issues early and improve code reliability.
Overall it will improve code maintainability.

tzuennn committed:
Refactor code to follow Single Level of Abstraction Principle (SLAP)

Break down the `parse` method in `Parser` class and `initializeComponents`
in `ChatUI` into smaller, focused methods. Extract initialization of
`ScrollPane` and `InputField` into separate methods.

This refactoring improves readability and maintainability by ensuring
each method has a single level of abstraction. It enhances code clarity
and simplifies future maintenance and extension.

tzuennn committed:
Merge pull request #1 from tzuennn/branch-A-Assertions

Add assertions in ChatUI and Parser classes

tzuennn committed:
Merge branch-A-CodeQuality to master

tzuennn committed:
Sort and display list of tasks by their dates

tzuennn committed:
Merge branch 'branch-C-Sort'

tzuennn committed:
Merge pull request #3 from tzuennn/branch-C-Sort

Sort and display list of tasks by their dates

tzuennn committed:
Merge branch 'master' of github.com:tzuennn/ip

tzuennn committed:
Move changes from master to branch-fxml

tzuennn committed:
Abstract Launcher to separate class

tzuennn committed:
Set cursor to be in textfield after submit button is clicked

tzuennn committed:
Refactor input handling and UI with FXML

Transitioned from Java code-based UI to FXML to enhance
the clarity and maintainability of the user interface.

The previous implementation used Java for constructing the UI,
which resulted in complex and hard-to-manage code. By adopting
FXML, the UI layout is now more straightforward and can be
easily modified without altering the underlying logic.

Additionally, I implemented more specific exception handling
for user input errors, particularly for incorrect date formats
and invalid command structures. This ensures that users receive
clear feedback when their input does not conform to the expected
formats, improving the overall user experience and robustness
of the application.

For example, input like `deadline homework by 2024-02-22 1900`
will now prompt the user with a specific error message,
indicating that the format is incorrect.

These enhancements contribute to cleaner code and a more
user-friendly interface, setting the stage for future
improvements and features.

tzuennn committed:
Update README.md to display user guide

tzuennn committed:
Merge branch 'branch-a-userguide'

tzuennn committed:
Fix style issues and update Javadoc

The codebase has several style issues that hinder readability and
maintainability, including long methods and lack of proper documentation.

Improving code readability and adherence to best practices enhances
maintainability.

What Is Being Done About It:
- Refactored methods in Nugget, Parser, TaskList, and Todo classes
  to follow the Single Level of Abstraction Principle (SLAP).
- Updated Javadoc comments for better clarity and documentation.
- Added a new UI diagram to the documentation to improve
  understanding of the user interface layout.

Refactoring ensures that methods remain focused and easier to
understand, while updated Javadoc provides clear guidance for future
developers.

This commit aims to improve overall code quality and documentation
in the project.

tzuennn committed:
Refactor message box UI to use fxml

tzuennn committed:
Update README.md
